### PR TITLE
Feature: Auth0 integration

### PR DESCRIPTION
**Pull request by xenirio**

### CODE REVIEW

File: app/api/v2/identity/sessions.rb [hunk 1]
```diff
@@ -87,6 +96,55 @@ class Sessions < Grape::API
...
+              response_type: 'code',
+              client_id: Barong::App.config.client_id,
+              redirect_uri: Barong::App.config.redirect_uri,
+              scope: 'openid profile email',
+              audience: Barong::App.config.audience
+            }.to_query)
+            redirect uri
+          end
+
+          params do
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: please, do description for every endpoint with success and failure

File: app/api/v2/identity/sessions.rb [hunk 2]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+
+          get '/authorize' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/authorize', query: {
+              response_type: 'code',
+              client_id: Barong::App.config.client_id,
+              redirect_uri: Barong::App.config.redirect_uri,
+              scope: 'openid profile email',
+              audience: Barong::App.config.audience
+            }.to_query)
+            redirect uri
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: why we should have redirect here?
I thought that we return auth token for FE, and FE will do this redirect with info which we send

File: app/api/v2/identity/sessions.rb [hunk 3]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+
+              present user, with: API::V2::Entities::UserWithFullInfo, csrf_token: csrf_token
+              status(200)
+            end
+            status(response.status)
+          end
+
+          get '/logout' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/v2/logout', query: {
+              client_id: Barong::App.config.client_id,
```
Comment (calj) [inline app/api/v2/identity/sessions.rb]: one param is snake case and the other camel case, really?

File: app/api/v2/identity/sessions.rb [hunk 4]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+              audience: Barong::App.config.audience
+            }.to_query)
+            redirect uri
+          end
+
+          params do
+            requires :code
+          end
+          get '/auth' do
+            url = "https://#{Barong::App.config.auth0_tenant_address}/oauth/token"
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: there is quite a lot logic here
Maybe lets do the auth0 helper module

File: config/initializers/barong.rb [hunk 5]
```diff
@@ -112,6 +116,14 @@
   config.set(:kycaid_authorization_token, '')
   config.set(:kycaid_sandbox_mode, 'true', type: :bool)
   config.set(:kycaid_api_endpoint, 'https://api.kycaid.com/')
+
+  # Auth0 configuration -----------------------------------------------
+  config.set(:auth0_tenant_address, 'openware-barong.us.auth0.com')
+  config.set(:client_id, ENV['CLIENT_ID'])
```
Comment (calj) [inline config/initializers/barong.rb]: here you need to put the default values (empty strings)

File: config/initializers/barong.rb [hunk 6]
```diff
@@ -38,6 +38,10 @@
 # Define default value for secret_key_base in test and development mode
 ENV['SECRET_KEY_BASE'] = '' unless Rails.env.production?
 
+# Define default value for Auth0
+ENV['CLIENT_ID'] ||= ''
```
Comment (calj) [inline config/initializers/barong.rb]: the default is set below

File: app/api/v2/identity/sessions.rb [hunk 7]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+              publish_session_create(user)
+
+              present user, with: API::V2::Entities::UserWithFullInfo, csrf_token: csrf_token
+              status(200)
+            end
+            status(response.status)
+          end
+
+          get '/logout' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/v2/logout', query: {
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: yeah, lets do auth0 helper module, with kind of login/logout things

File: config/initializers/barong.rb (hunk 5 above)
Comment (calj) [inline config/initializers/barong.rb]: We should prefix those variables.
If this is specific to Auth0 let's prefix with AUTH0
Is this code compatible with OAuth protocol? In this case we would prefix with OAUTH.

**Review by calj — CHANGES_REQUESTED** [review]
Please add documentation

File: app/api/v2/identity/sessions.rb [hunk 8]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+                 { code: 400, message: 'Required params are empty' },
+                 { code: 404, message: 'Record is not found' }
+               ]
+
+          get '/authorize' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/authorize', query: {
+              response_type: 'code',
+              client_id: Barong::App.config.client_id,
+              redirect_uri: Barong::App.config.redirect_uri,
+              scope: 'openid profile email',
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: which scopes we can put here?
If its kind of customization, maybe we should do this with env too

File: app/api/v2/identity/sessions.rb [hunk 9]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+              req.body = URI.encode_www_form(data)
+            end
+            if response.status == 200
+              json = JSON.parse(response.body)
+              id_token = json['id_token']
+
+              claims = JWT.decode(id_token, nil, false, algorithm: 'RS256')
+              user = get_user(claims[0]['email'])
+
+              activity_record(user: user.id, action: 'login::2fa', result: 'succeed', topic: 'session')
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: why action with login::2fa ?

File: app/api/v2/identity/sessions.rb [hunk 10]
```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
...
+            }
+            response = Faraday.post(url) do |req|
+              req.headers['Content-Type'] = 'application/x-www-form-urlencoded'
+              req.body = URI.encode_www_form(data)
+            end
+            if response.status == 200
+              json = JSON.parse(response.body)
+              id_token = json['id_token']
+
+              claims = JWT.decode(id_token, nil, false, algorithm: 'RS256')
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: check if claims has email (to prevent nil cases)

**Review by chumaknadya — CHANGES_REQUESTED** [review]

File: app/api/v2/identity/sessions.rb (hunk 3 above)
Comment (xenirio) [reply app/api/v2/identity/sessions.rb]: Correct, From auth0 Website. :D
https://auth0.com/docs/logout/redirect-users-after-logout

Screen Shot 2564-01-26 at 15 20 54

File: docs/auth0.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+Barong provide the endpoints support to create the barong session with auth0 `id_token`
+
+### Credentials configuration
+##### This document provides description and defaults for all environment variables that take a part in app configuration
+| Env name | Default value | Description |
+| ---------- | ------ | ---------------------------------- |
+| `barong_auth0_tenant_address` | | Auth0 application domain |
+| `barong_auth0_client_id` | | Auth0 application client ID |
+| `barong_auth0_client_secret` | | Auth0 application secret |
+| `barong_auth0_audience` | | Unique identifier for the API. This value will be used as the `audience` parameter on authorization calls |
```
Comment (chumaknadya) [inline docs/auth0.md]: this should be in configuration.md

File: app/api/v2/identity/sessions.rb [hunk 12]
```diff
@@ -87,6 +93,41 @@ class Sessions < Grape::API
...
+          end
+
+          desc 'Logout from Auth0',
+               success: { code: 302, message: 'User logged out' },
+               failure: [
+                 { code: 404, message: 'Record is not found' }
+               ]
+          get '/logout' do
+            uri = Barong::Auth0.new.logout_uri(Barong::App.config.auth0_client_id, Barong::App.config.auth0_logout_uri)
+            redirect uri
```
Comment (chumaknadya) [inline app/api/v2/identity/sessions.rb]: we should destroy session in barong too

and don't need redirect

Comment (xenirio) [reply app/api/v2/identity/sessions.rb]: I think we don't need the logout form auth0 in Barong. The logout should belong to FE. Just session destroy is enough. I will remove this endpoint.

File: docs/general/auth0.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
+# Auth0 integration
+
+## How to create application
```
Comment (Ievgenii78) [inline docs/general/auth0.md]: an application

File: docs/general/auth0.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+When you signed up for Auth0, a new application was created for you, or you could have created a new one (the most appropriate application for our structure is Single Page Application).
+![Application](../images/auth0_dashboard.png)
+
+You will need some details about that application to communicate with Auth0. You can get these details from the Application Settings section in the Auth0 dashboard.
+
+![Settings](../images/auth0_settings.png)
+
+You should put `auth0_domain` from Domain field and `auth0_client_id` from Client ID field
```
Comment (Ievgenii78) [inline docs/general/auth0.md]: from the Domain
from the Client

File: docs/general/auth0.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+For single page application better to use authorization code flow with proof key for code exchange.
```
Comment (Ievgenii78) [inline docs/general/auth0.md]: For a single page
a proof

**Pull request closed without merging**